### ISSUE TITLE
Metronome GC - pause times increase with uptime

### ISSUE THREAD

**Issue by akrus** [eclipse/openj9#4769]
Hello!

I noticed strange behavior when OpenJ9 is running when application is running as a daemon.

![image](https://user-images.githubusercontent.com/36325/52934824-40e2e780-3360-11e9-9486-455a3e962d2a.png)
As you can see, 'GC quantum time maximum' usually is at 5 msec when the application is just started, increasing to 8-10 msec after ~40 hours of runtime (this varies and can be in the range 20-50 hours), but after ~105 hours it increases to 20 msec. I was unable to identify any relations between workload and pause times, but on another server I had an increase to 70-80 msec.

See attached verbose GC log and TGC rootscantime log.
[gc_tgc.zip](https://github.com/eclipse/openj9/files/2874455/gc_tgc.zip)

OS version: Ubuntu 18.04, kernel 4.15.0-38-lowlatency
JVM version (I also experience same problem with 0.9.0 and 0.11.0):
```
openjdk version "1.8.0_202"                                                                                                                                                                                                                  
OpenJDK Runtime Environment (build 1.8.0_202-b08)                                                                                                                                                                                            
Eclipse OpenJ9 VM (build openj9-0.12.1, JRE 1.8.0 Linux amd64-64-Bit 20190205_196 (JIT enabled, AOT enabled)
OpenJ9   - 90dd8cb40
OMR      - d2f4534b
JCL      - d002501a90 based on jdk8u202-b08)                                                                                                                                                                                                 
```

This was briefly discussed with @amicic before.

**Comment by amicic** [comment]
The logs indicate that scanning finalization queue (of ready to be finalized, but unfinalized yet) of objects is a problem.

Here is a sample log:

<scan timestamp="Feb 15 07:37:55 2019">
	<thread id="0" ownablesynchronizerobjects="0.001" phantomrefscomplete="0.125" unfinalizedobjectscomplete="0.192" maxincrementtime="0.125" maxincremententity="unfinalizedobjectscomplete"/>
	<thread id="1" ownablesynchronizerobjects="0.002" phantomrefscomplete="0.092" unfinalizedobjectscomplete="13.424" maxincrementtime="13.387" maxincremententity="unfinalizedobjectscomplete"/>
	<thread id="2" ownablesynchronizerobjects="0.002" phantomrefscomplete="0.095" unfinalizedobjectscomplete="13.431" maxincrementtime="13.393" maxincremententity="unfinalizedobjectscomplete"/>

**Comment by akrus** [comment]
@amicic, is there anything else I can collect from my side?

**Comment by amicic** [comment]
@akrus, I analyzed the TGC logs more deeply. Unfortunately, they did not point to a real problem in root scanning. It's possible that there is no problem in root scanning at all, but I also noticed some weaknesses in the tool.

I improved the log reporting tool further: https://github.com/eclipse/openj9/pull/4829 and https://github.com/eclipse/openj9/pull/4478. So you'll either wait for 0.13 release (in about a month), or if you can reproduce the problem in non-production environment, you can try a nightly build (in a day or two).

**Comment by akrus** [comment]
@amicic, OK, I'll try to reproduce this with a nightly build. Thanks!

**Comment by amicic** [comment]
ok I believe you can watch builds here: https://ci.eclipse.org/openj9/view/Build/job/Build-JDK8-linux_x86-64/
look for 'Tgc root scan for Metronome' in Nightly type of builds (not OMR-Acceptance type)

**Comment by pshipton** [comment]
Note the 0.13 release will only include Java 12 builds.

**Comment by pshipton** [comment]
Preferably nightly builds from AdoptOpenJDK.net should be used. The builds we do at the OpenJ9 project for test purposes may not contain everything you get in a build from Adopt.

https://adoptopenjdk.net/nightly.html?variant=openjdk8&jvmVariant=openj9

**Comment by akrus** [comment]
I was just checking another server and noticed a strange thing:
![image](https://user-images.githubusercontent.com/36325/53263527-5189b980-36e2-11e9-9de3-5294a07dcab3.png)
As you can see 'free heap' is constantly decreasing.
Here is max quantum times:
![image](https://user-images.githubusercontent.com/36325/53263564-6d8d5b00-36e2-11e9-8885-3c002960ea6c.png)

Can it be a memory leak?

**Comment by akrus** [comment]
@amicic, max quantum time didn't increase that much yet on the 2019-02-25 build, but still higher than initially.

[logs_20190227.tar.gz](https://github.com/eclipse/openj9/files/2910831/logs_20190227.tar.gz)

**Comment by akrus** [comment]
Updated log, now more clearly visible.
[logs_20190228.tar.gz](https://github.com/eclipse/openj9/files/2915271/logs_20190228.tar.gz)

Now maxincrementtime is here:
```
<scan timestamp="Feb 28 13:43:29 2019">
        <thread id="0" classloaders="0.109" threads="7.285" jniglobalrefs="0.003" maxincrementtime="2.552" maxincremententity="threads"/>
        <thread id="1" threads="3.229" maxincrementtime="0.974" maxincremententity="threads"/>
        <thread id="2" threads="3.230" maxincrementtime="0.979" maxincremententity="threads"/>
        <thread id="3" threads="3.249" maxincrementtime="0.978" maxincremententity="threads"/>
        <thread id="4" threads="3.222" maxincrementtime="0.977" maxincremententity="threads"/>
        <thread id="5" threads="3.226" maxincrementtime="0.976" maxincremententity="threads"/>
        <thread id="6" threads="3.233" maxincrementtime="0.975" maxincremententity="threads"/>
        <thread id="7" threads="3.065" maxincrementtime="0.968" maxincremententity="threads"/>
        <thread id="8" threads="3.225" maxincrementtime="0.975" maxincremententity="threads"/>
        <thread id="9" threads="3.228" maxincrementtime="0.973" maxincremententity="threads"/>
        <thread id="10" threads="3.204" maxincrementtime="0.970" maxincremententity="threads"/>
        <thread id="11" threads="3.236" maxincrementtime="0.976" maxincremententity="threads"/>
        <total classloaders="0.109" threads="42.632" jniglobalrefs="0.003"/>
</scan>
```

**Comment by amicic** [comment]
Hi, @akrus I have a through look yesterday (at logs from 27th). I belive that root scanning stats now report accurate date. However, they do not detect any pauses longer than 2-3ms in roots (which is occasionally a scan of thread stack).

At this moment I do not have any theory where the pause could be coming from. 

Btw, root scanning was a suspect from the very beginning since it's what we do at the start and the end of a cycle. And the 10+ms pause that verbose GC reports was always in the first heart beat.

Could be that there is some pause even before root scanning (there should not be much activity before root scanning except perhaps mark map related), or some transition between root scanning and main recursive scan phase....  I'll probably add some extra logging (unrelated to root scanning).

One observation is that the increase in pause times often occur in not so small increments. There are two of them visible here: 
![metronome_max_pause](https://user-images.githubusercontent.com/18264984/53594577-26cbb500-3b69-11e9-86ed-b1a44f016b7a.jpg)

Also a question.... VM appear to run with 12 GC threads, while there are 24 threads available (which should normally default to as many GC threads). Is this machine with 2 NUMA nodes? Is JVM pinned to one of them?

**Comment by akrus** [comment]
Yes, it's 2 NUMA nodes and JVM is pinned.

**Comment by amicic** [comment]
Please generate javacore (kill -3 `pid of JVM`) at any point once there are 10+ms observed.

**Comment by amicic** [comment]
Regarding NUMA.... Your heap is 200G and machine appear to have have around 370G. Hence the heap at one point many not be commited from the requested NUMA node. 

You should monitor cat /proc/`jvm PID`/numa_maps and see if occurrence of increase of pauses  coincide with increase of heap commited region (you can eyeball heap as being the region with highest number of pages commited). If you for example bind to N0, and you start seeing the region having (increase) of N1 pages. 
Or simple try to reduce the heap size to about 170G, so that it fits into one node.

**Comment by akrus** [comment]
Yes, I was just playing with the env when we had previous problem and increased it from 175g. It's still reproducible in both cases, but I'll have a look on NUMA stats.

**Comment by akrus** [comment]
@amicic, got a javacore, what should I do with it?

**Comment by amicic** [comment]
You can attach javacore here or just send me to my business email which I believe you already know. Interesting section is about duration of various hooks (not that really suspect hooks are problem, but it's worthwhile checking).

Meanwhile I modified slightly  how we report the timestamp for max pause within a heartbeat. Please generate a new pair or verbose gc / tgc logs, using a fresh nightly build.

Anything interesting from numa maps?

**Comment by akrus** [comment]
OK, sent javacore and updated OpenJ9 to 2019-03-06 build from AdoptOpenJDK.

NUMA is OK.

**Comment by amicic** [comment]
Hooks info somewhat suggests that verbose gc logging itself may be a problem. I/O operation (output of verbose GC text data to a file) could block a GC thread, and delay the finish of a GC increment. 

Next time please try with option -Xgc:bufferedLogging. Note that the output will be delayed (intermediate buffer is flushed by a dedicated logging thread, when full), which could be arbitrary delay in time terms.  Also, remove TGC logging, because this buffered logging applies only to verbose GC.

**Comment by akrus** [comment]
OK, added this flag and disabled TGC log.

**Comment by akrus** [comment]
-Xgc:bufferedLogging didn't change anything, max quantum time increased again.

**Comment by amicic** [comment]
I have looked at latest logs. The timestamp of max pause is always 0ms (and we already knew the long pause was in first 1sec verbose gc heartbeat). For example:

<trigger-start id="5735" timestamp="2019-03-08T07:17:30.027" />
<cycle-start id="5736" type="global" contextid="0" timestamp="2019-03-08T07:17:30.029" intervalms="124585.442" />
<gc-op id="5737" type="heartbeat" contextid="5736" timestamp="2019-03-08T07:17:31.033">
  <quanta quantumCount="168" quantumType="mark" minTimeMs="0.824" meanTimeMs="1.112" maxTimeMs="8.581" **maxTimestampMs="0.000"** /

It means the long pause is always within the very first GC increment of a GC cycle. Within a cycle, real GC work starts with root scanning, and since TGC root scanning logs don't report any long pauses, seems like this pause is within preparation steps. Those are really simple things that we do, checking/setting some flags etc.. I checked the code again and I don't see anything computationally intensive (like iteration through all regions etc). 

I'll probably add some more tracing/logging to break down the preparation steps and narrow down possible suspects.

**Comment by amicic** [comment]
I apologize... I did not have time to work on this (to introduce new trace points to narrow down where the delay is). But they are a couple of already that could be interesting, although chances are I'll need to introduce new ones. Could you run with:

-Xtrace:none,maximal=j9mm.54,maximal={j9mm.190-191},maximal={j9mm.474-475},maximal={j9mm.467-468},maximal=j9mm.630,output="output.trc"

If you have 2 or more JVMs make sure that output file name is different (for example, use output%p.trc) or have the logs in different dirs....

Files will probably be a few times bigger than verbose GC. 
I need matching verbose GC files as well.

**Comment by akrus** [comment]
@amicic, not sure what's wrong, but it complains:
JVMTRCE004W Error opening trace file: "/home/log/jvm_20190328075145_2073.log.trc"

**Comment by amicic** [comment]
An excerpt from trace point logs:

06:16:11.027736000 0x00007f4690003c00 j9mm.467 Event Cycle Start: type 1 approximateFreeMemorySize 93952224376
06:16:11.027773000 0x00007f4690003c00 j9mm.474 Event GlobalGC start: globalcount=554
06:16:11.036980000 0x00007f4690003c00 j9mm.190 Entry >MM_GlobalCollector_isTimeForClassUnloading. dynamicClassUnloading:1 classLoaderBlocks:381 dynamicClassUnloadingThreshold:1 lastUnloadNumOfClassLoaders:381

As before, this shows that the pause very early in a GC cycle before we even start heavy GC work.

**Comment by amicic** [comment]
Introducing new trace points to narrow down the root cause: https://github.com/eclipse/omr/pull/3746

**Comment by amicic** [comment]
From the latest logs, the pause appear to come between these two trace points:

06:19:38.596243000  0x00007ef7f4003c00 j9mm.707            ERROR     This tracepoint's format string was not available in dat file.
06:19:38.605757000  0x00007ef7f4003c00 j9mm.709            ERROR     This tracepoint's format string was not available in dat file.

For some reason, the formatter did not recognize the new trace points (perhaps akrus used an older JVM to format them?). Anyhow, from the IDs, looking into the matching DAT file, I can see the points in questions are:

j9mm.707 4 1 2 N Trc_MM_RealtimeGC_internalPreCollect_Exit " <RealtimeGC::internalPreCollect"^M
j9mm.709 4 1 2 N Trc_MM_Collector_preCollect_Exit " <Collector::preCollect"^M

Looking into the source code, the only non-trivial computation comes from gathering stats for Excessive GC:

https://github.com/amicic/omr/blob/metronome_cycle_start_tracing/gc/base/Collector.cpp#L254

Excessive GC computation makes little sense for this type of collector, and it can probably be disabled. 

Still, what this does is gather free memory stats. The most likely offender is counting 'single' free regions (by iterating the single free list of Region descriptors). This list gets larger over time, as spans of free contiguous regions get broken down into smaller spans, and eventually into single regions.

In these runs with 175G heaps, there are about 2.7M regions. This iteration occurs at the beginning of a cycle, when heap is ~50% free, so it could be up to ~1.3M regions in the free list. Does not sound too much (to cause ~10ms delay), but  if we account for cache misses when visiting all those Region Descriptors it sound more plausible.

A possible fix could go along caching the region count and maintaining it on every enqueue/dequeue, which would avoid iterating the list every time we need it's current size.

A workaround is to increase the region size, what will decrease the list size. This has been already discussed here https://www.eclipse.org/lists/openj9-dev/msg00040.html, where I (randomly) recommended 4x larger than default, but I guess at that time (with relatively short runs) there were no visible benefits. 

@akrus Long story short, please try larger region size. Knowing that this computation step takes ~10ms I'd say 8x larger than default (8x65536=524288) is a reasonable starting point. And if you see that it helps and creates no negative side effect, then even larger  region could be tried. Theoretically, a negative side effect is increase in heap fragmentation, which would make GC more frequent, but it's hard to estimate if that will have any tangible  effect in practice. 

While trying larger region size, generating same trace logs would be wise.

Meanwhile, I'll think of a fix.

**Comment by amicic** [comment]
While we have an option to disable excessive GC logic, it will not help here, because the stats are gathered regardless. Possibly another thing we could change.

**Comment by akrus** [comment]
@amicic, sorry, indeed used an older build to generate formatted trace, sent you an updated one (though you likely do not need it anyway now). I've changed region size to 524288, so probably beginning of next week I'll have new logs.

Although probably it's worth mentioning that I noticed the same behaviour on smaller heaps as well (~20G), but it's less noticeable. Also it was much more easier to reproduce when app was leaking memory heavily.

**Comment by akrus** [comment]
@amicic, I've tried changing region size to 524288 and it looks better now, I see an average decrease of 0.5 ms on max quantum size and generally looks like GC is running less frequently. Should I try changing region size further? Considering huge pages are on and page size is 2M, is it worth trying to set region size to 2M as well? Or it's not going to be aligned properly anyway?

**Comment by amicic** [comment]
Here is a sample cycle before:
```
<cycle-start id="8431" type="global" contextid="0" timestamp="2019-03-29T15:23:36.721" intervalms="124116.524" />
  <quanta quantumCount="106" quantumType="mark" minTimeMs="0.916" meanTimeMs="1.186" maxTimeMs="12.149" maxTimestampMs="112797717.096" />
  <quanta quantumCount="127" quantumType="mark" minTimeMs="1.032" meanTimeMs="1.077" maxTimeMs="1.172" maxTimestampMs="112799007.984" />
  <quanta quantumCount="154" quantumType="mark" minTimeMs="1.028" meanTimeMs="1.078" maxTimeMs="1.206" maxTimestampMs="112800243.993" />
  <quanta quantumCount="180" quantumType="mark" minTimeMs="0.809" meanTimeMs="1.077" maxTimeMs="1.180" maxTimestampMs="112801010.245" />
  <quanta quantumCount="123" quantumType="mark" minTimeMs="1.034" meanTimeMs="1.083" maxTimeMs="1.246" maxTimestampMs="112802755.187" />
  <quanta quantumCount="3" quantumType="mark" minTimeMs="0.210" meanTimeMs="0.762" maxTimeMs="1.040" maxTimestampMs="112802771.099" />
  <quanta quantumCount="199" quantumType="sweep" minTimeMs="0.826" meanTimeMs="1.033" maxTimeMs="4.699" maxTimestampMs="112802779.331" />
  <quanta quantumCount="171" quantumType="sweep" minTimeMs="0.726" meanTimeMs="1.004" maxTimeMs="1.070" maxTimestampMs="112804021.109" />
  <quanta quantumCount="182" quantumType="sweep" minTimeMs="0.975" meanTimeMs="1.015" maxTimeMs="1.062" maxTimestampMs="112805128.043" />
  <quanta quantumCount="198" quantumType="sweep" minTimeMs="0.888" meanTimeMs="0.957" maxTimeMs="1.040" maxTimestampMs="112805798.674" />
  <quanta quantumCount="146" quantumType="sweep" minTimeMs="0.766" meanTimeMs="0.984" maxTimeMs="2.812" maxTimestampMs="112807594.583" />
  <quanta quantumCount="5" quantumType="postcollect" minTimeMs="0.474" meanTimeMs="0.863" maxTimeMs="1.028" maxTimestampMs="112807609.577" />
<cycle-end id="8444" type="global" contextid="8431" timestamp="2019-03-29T15:23:46.631" />
```
and after large region size tuning:
```
<cycle-start id="8115" type="global" contextid="0" timestamp="2019-04-25T06:32:05.475" intervalms="204486.425" />
  <quanta quantumCount="170" quantumType="mark" minTimeMs="1.006" meanTimeMs="1.100" maxTimeMs="1.609" maxTimestampMs="0.000" />
  <quanta quantumCount="155" quantumType="mark" minTimeMs="1.022" meanTimeMs="1.085" maxTimeMs="1.211" maxTimestampMs="175.328" />
  <quanta quantumCount="193" quantumType="mark" minTimeMs="0.603" meanTimeMs="1.067" maxTimeMs="1.219" maxTimestampMs="1006.300" />
  <quanta quantumCount="179" quantumType="mark" minTimeMs="0.245" meanTimeMs="1.088" maxTimeMs="1.210" maxTimestampMs="164.778" />
  <quanta quantumCount="188" quantumType="sweep" minTimeMs="0.771" meanTimeMs="1.051" maxTimeMs="5.272" maxTimestampMs="0.000" />
  <quanta quantumCount="199" quantumType="sweep" minTimeMs="0.800" meanTimeMs="1.025" maxTimeMs="1.067" maxTimestampMs="461.437" />
  <quanta quantumCount="72" quantumType="sweep" minTimeMs="0.753" meanTimeMs="1.009" maxTimeMs="3.154" maxTimestampMs="331.711" />
  <quanta quantumCount="5" quantumType="postcollect" minTimeMs="0.394" meanTimeMs="0.898" maxTimeMs="1.029" maxTimestampMs="0.000" />
<cycle-end id="8124" type="global" contextid="8115" timestamp="2019-04-25T06:32:11.846" />
```

**Comment by amicic** [comment]
Clearly, the long pause time in the first quanta is gone. This specific problem was investigated in this issue, and seems to be resolved by tuning. As I mentioned before, we could possibly come up with a fix too (too busy with other stuff, so I'll probably delegate someone to work on a fix).

Another possible (positive) side effect is shorter GC cycle, in particular sweep phase (5 vs 3 quantas). I did not see any negative side effects.

However, there is a secondary problem that I ignored all along, and those are pauses at the beginning and the end of sweep phase of about 5ms and 3ms respectively. The region size tuning did not seem affect them. This warrants whole new investigation. I would even suggest you open a different issue, and distinguish from this one in subject with something like 'long pause in sweep phase' (and of course mention Metronome), and possibly even rename subject of this one with something like 'long pause early in GC cycle'. 

As for you question, if to increase region size even further, I'd say no need. 2M large page should have little bearing. Btw, regions sizes are aligned too, so we will fit exactly 4 regions in one large page.

**Comment by amicic** [comment]
Actually, you don't need to rename this one. 'increase with uptime' makes this one fairly distinguishable.

**Comment by akrus** [comment]
@amicic, but for larger heaps of 500G+ is it better to increase the region size? Or 512k should be OK there also?

**Comment by amicic** [comment]
Hard to say, it's possibly you'll need to increase even further, probably double. You'll just have to try. Sorry, but I can't be more specific - we don't have experience with that large heaps using Metronome, hence we are seeing all these problems. 

Again, I'm afraid of negative side effects. If I remember correctly granularity of work in sweep phase is a region. With too larger region, sweep increments could start growing up (beyond what what we already see it).

**Comment by amicic** [comment]
@akrus We put a fix, so that no workaround (with larger region size) is required any more. 

Perhaps it's still good to keep region size slightly larger (128K or 256K) than default (64K)  to reduce metadata processing overhead (and shorten GC cycle time). rather than (512K or 1M), which could cause more fragmentation and longer pauses in sweep (since sweep use 1 region as unit of work, it may be >1ms to sweep a very larger region).

**Comment by akrus** [comment]
@amicic, thanks! I've updated the build here with AdoptOpenJDK nightly from 02/06/2019, so will probably have some results in a couple of days.

**Comment by akrus** [comment]
@amicic, I can confirm that fix works well and there is no noticeable increase in GC pause times even with default (64k) region size. I'll try increasing to 128k to check if there is any difference. But I think this bug report can be closed as #5615 is there.
